src/util/address.rs

```diff
@@ -225,48 +299,89 @@ impl Display for Address {
+                    Network::Regtest => "bcrt",
+                };
+                bech32::encode_to_fmt(fmt, &hrp, &b32_data).expect("only errors on invalid HRP")
+            }
         }
     }
 }
 
+/// Extract the bech32 prefix.
+/// Returns the same slice when no prefix is found.
+fn find_bech32_prefix(bech32: &str) -> &str {
+    // Split at the last occurrence of the separator character '1'.
+    match bech32.rfind("1") {
+        None => bech32,
+        Some(sep) => bech32.split_at(sep).0,
+    }
+}
+
 impl FromStr for Address {
-    type Err = encode::Error;
-
-    fn from_str(s: &str) -> Result<Address, encode::Error> {
-        // bech32 (note that upper or lowercase is allowed but NOT mixed case)
-        if s.starts_with("bc1") || s.starts_with("BC1") ||
-           s.starts_with("tb1") || s.starts_with("TB1") ||
-           s.starts_with("bcrt1") || s.starts_with("BCRT1")
-        {
-            let witprog = WitnessProgram::from_address(s)?;
-            let network = match witprog.network() {
-                bitcoin_bech32::constants::Network::Bitcoin => Network::Bitcoin,
-                bitcoin_bech32::constants::Network::Testnet => Network::Testnet,
-                bitcoin_bech32::constants::Network::Regtest => Network::Regtest,
-                _ => panic!("unknown network")
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Address, Error> {
+        // try bech32
+        let bech32_network = match find_bech32_prefix(s) {
+            // note that upper or lowercase is allowed but NOT mixed case
+            "bc" | "BC" => Some(Network::Bitcoin),
+            "tb" | "TB" => Some(Network::Testnet),
+            "bcrt" | "BCRT" => Some(Network::Regtest),
+            _ => None,
+        };
+        if let Some(network) = bech32_network {
+            // decode as bech32
+            let (_, payload) = bech32::decode(s)?;
+            if payload.len() == 0 {
+                return Err(Error::EmptyBech32Payload);
+            }
+
+            // Get the script version and program (converted from 5-bit to 8-bit)
+            let (version, program) = {
+                let (v, p5) = payload.split_at(1);
+                (v[0], Vec::from_base32(p5)?)
             };
-            if witprog.version().to_u8() != 0 {
-                return Err(encode::Error::UnsupportedWitnessVersion(witprog.version().to_u8()));
+
+            // Generic segwit checks.
+            if version.to_u8() > 16 {
+                return Err(Error::InvalidWitnessVersion(version.to_u8()));
             }
+            if program.len() < 2 || program.len() > 40 {
+                println!("xxxx");
+                return Err(Error::InvalidWitnessProgramLength(program.len()));
+            }
+
+            // Specific segwit v0 check.
+            if version.to_u8() == 0 && (program.len() != 20 && program.len() != 32) {
+                return Err(Error::InvalidSegwitV0ProgramLength(program.len()));
```

Comment:
Is this actually an error? I thought it would just be nonstandard?

Reply:
Well, I'd prefer to have it in as an anti-footgun, no? Given latest protocol rules, such addresses are blackholes and will always be blackholes, right?

Reply:
>  If the version byte is 0, but the witness program is neither 20 nor 32 bytes, the script must fail.

https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#witness-program

Reply:
Oh ok, sounds good. Agreed we don't want to support such "addresses"

Reply:
If you really feel strongly, I could add a `from_str_unchecked` or something.